services/link-preview/LinkPreviewService.ts

```diff
@@ -0,0 +1,222 @@
+
+        if (contentType.startsWith('image/')) {
+            const hostname = new URL(url).hostname;
+
+            const result = await this.getLinkPreviewImageData(proxiedUrl);
+
+            linkPreview = {
+                imageUrl: proxiedUrl,
+                imageData: result.data,
+                imageContentType: result.contentType,
+                domainUrl: hostname,
+            };
+        } else if (contentType.startsWith('text/')) {
+            linkPreview = await this.getLinkPreviewFromResponseData({
+                url: url,
+                htmlContent: response.data
+            });
+        }
+
+        this.cache.set(url, linkPreview);
+
+        return linkPreview;
+    }
+
+    private async getLinkPreviewFromResponseData({url, htmlContent}: {url: string, htmlContent: string}): Promise<LinkPreview> {
+        const linkPreview: LinkPreview = {};
+
+        const parser = new DOMParser();
+        const dom = parser.parseFromString(htmlContent, 'text/html');
+
+        const parsedURL = new URL(url);
+
+        linkPreview.domainUrl = parsedURL.hostname;
+        linkPreview.faviconUrl = this.getFaviconFromDocument({
+            dom,
+        });
+
+        const openGraphMetaTags = dom.head.querySelectorAll('[property*="og:"');
+        openGraphMetaTags.forEach(tag => {
+            const type: string = tag.attributes['property']?.value.split('og:').pop();
+            const value: string = tag.attributes['content']?.value;
+
+            switch (type.toLowerCase()) {
+                case 'title':
+                    linkPreview.title = value;
+                    break;
+                case 'description':
+                    linkPreview.description = value;
+                    break;
+                case 'image':
+                    linkPreview.imageUrl = value;
+                    break;
+                case 'site_name':
+                    linkPreview.siteName = value;
+                    break;
+            }
+        });
+
+        if (!linkPreview.title) {
+            // fallback
+            const titleElement = dom.head.querySelector('title');
+            if (titleElement && titleElement.innerText.length) {
+                linkPreview.title = titleElement.innerText;
+            } else {
+                // show the domain as final fallback
+                linkPreview.title = linkPreview.domainUrl;
+            }
+        }
+
+        if (!linkPreview.imageUrl) {
+            // fallback
+            const images = dom.querySelectorAll('img');
+            if (images && images.length) {
+                for (const image in images) {
+                    const imageElement = images[image];
+                    const imageSrc: string = imageElement?.attributes?.['src']?.value;
+
+                    if (!imageSrc) {
+                        continue;
+                    }
+
+                    if (imageSrc.endsWith('jpg') || imageSrc.endsWith('gif') || imageSrc.endsWith('png')) {
```

Comment:
Maybe add SVG too.

Reply:
I thought about this but I wasn't entirely sure if SVGs would be safe enough. But as long as they're inside of an img tag and base64 encoded it shouldn't be an issue I think.